Fix categorical (factor) handling in native R reader

jackkamm opened the pull request:
When I try to use the native R reader on an h5ad file I have, like so:

```
sce <- readH5AD("/path/to/my/anndata.h5ad", 
                reader='R',
                use_hdf5=TRUE,
                verbose=TRUE)
```

I get the following warning:
```
Warning messages:
1: In value[[3L]](cond) : setting 'colData' failed for
  '/path/to/my/anndata.h5ad':
  cannot coerce class "list" to a DataFrame
2: In value[[3L]](cond) : setting 'rowData' failed for
  '/path/to/my/anndata.h5ad':
  cannot coerce class "list" to a DataFrame
```

And the `colData` and `rowData` of the returned SCE are empty.

Tracking it down, it's because zellkonverter was unable to convert some Categorical (factor) columns in the obs/var. It's because AnnData 0.8 changed the way Categoricals are encoded.

This updates the native R reader so it can read factors encoded with the newer format. After applying this patch I was able to read my h5ad file.

EDIT:

Related: https://github.com/theislab/zellkonverter/issues/78

Thought I should also add why I'm using the R reader rather than the recommended python reader. It's because the python reader had errors reading X/layers matrices from my h5ad file due to unsorted indptr (seems to be the same issue as https://github.com/theislab/anndata2ri/issues/51). I was able to fix some of the errors by calling `sort_indices()` and resaving the AnnData, but some matrices still wouldn't load, and the ones that did were loaded as dgCMatrix instead of DelayedArray. By contrast I found the R reader successfully read all matrices from my h5ad file as DelayedArray.

LTLA commented:
@lazappi will have more comments, but one thing I'll note is that the change must be back-compatible with the previous H5AD formats. I'm not familiar with the differences but it sounds like `__categories` no longer exists in the new format, in which case you could check for that in `names(fields)` and switch to your approach if it doesn't exist. This avoids weird interactions between old versions of the files that happen to use the new keywords.

(A better approach would be to detect the file version up-front and pass it along to each internal function, which avoids the need for guessing inside each function. However, this would involve a more dramatic set of changes.)

Also, I have to say it, but the surrounding code uses 4-space indenting, and so should you.

jackkamm commented:
> Also, I have to say it, but the surrounding code uses 4-space indenting, and so should you.

Thanks for catching that, fixed now.

> but one thing I'll note is that the change must be back-compatible with the previous H5AD formats

I think the changes should be back-compatible. I did not delete the handling for the old categories format, but simply added code to handle factors encoded in the new format. Theoretically I think it could even handle an h5ad that contains a mix of factors encoded in both styles (not that such a scenario would happen).

I could switch to an explicit if/else instead, but not yet sure the best way to check which format the AnnData is.

lazappi commented:
Hi @jackkamm. Thanks for the contribution! The R reader is still a work in progress but I know some people have tried using it so this will be appreciated. I just have a couple of points:

- There is now a written spec for the **anndata** 0.8 H5AD format https://anndata.readthedocs.io/en/latest/fileformat-prose.html. If you could check that everything is consistent that would be great.
- It would be great to have at least one test for this (either a normal test or a long test). The simplest thing might be to write a file with the Python 0.8 writer and see if it can be read with the R reader (making sure it has the right kind of columns).
- Do you think this solves #78 as well or is that a slightly different issue?

jackkamm commented:
> There is now a written spec for the anndata 0.8 H5AD format https://anndata.readthedocs.io/en/latest/fileformat-prose.html. If you could check that everything is consistent that would be great.

Thanks for the reference -- looks like more work needs to be done to support nullable booleans and ints as per the spec. I will look into it.

> It would be great to have at least one test for this (either a normal test or a long test). The simplest thing might be to write a file with the Python 0.8 writer and see if it can be read with the R reader (making sure it has the right kind of columns).

Sounds good, I've added a test, still a work-in-progress since it needs to check the nullable booleans mentioned above.

> Do you think this solves https://github.com/theislab/zellkonverter/issues/78 as well or is that a slightly different issue?

I think it's basically the same issue, I was getting the same error message, and the PR fixed it for me (as in I was able to read in the h5ad). But the original PR only fixed a subset of the problem (factors/categories), a little more work needs to be done to make sure other types are also converted correctly.

lazappi commented:
👍🏻 It sounds like you are still working on some things which is great, just ping me when you are ready for a review

jackkamm commented:
I've added handling for nullable ints and bools, so I believe the implementation satisfies the v0.8 spec now [1], and is ready for a review.

For the test, I found that `writeH5AD()` wasn't quite consistent with the spec [2], therefore I manually created a separate AnnData v0.8 object in Python [3], which I saved to the git repo.

Footnotes:

[1] The spec also mentions string handling, but I didn't test it because AnnData.write always converts strings to factors: https://github.com/scverse/anndata/blob/8e793af01a77d0e31e91a72f3988df7d6de9cdc5/anndata/_io/h5ad.py#L58

[2] `writeH5AD` seems to convert `NA_integer_` to `2^-31`, instead of using a mask as in the spec.

[3] Here is the Python code I used to create `krumsiek11_augmented_v0-8.h5ad`:

```
import numpy as np
import pandas as pd
import anndata as ad

adata = ad.read_h5ad("/home/jack/src/zellkonverter/inst/extdata/krumsiek11.h5ad")

# add float column to colData/obs
adata.obs['dummy_num'] = 42.42

# float column with NA
adata.obs['dummy_num2'] = adata.obs['dummy_num']
adata.obs['dummy_num2'][0] = float('nan')

# int column
adata.obs['dummy_int'] = np.arange(adata.shape[0])

# int column with NA
adata.obs['dummy_int2'] = pd.array([None] + [42] * (adata.shape[0] - 1))

# bool column
adata.obs['dummy_bool'] = True
adata.obs['dummy_bool'][0] = False

# bool column with NA
adata.obs['dummy_bool2'] = pd.array([False, None] + [True] * (adata.shape[0] - 2))

# also add some entries to the metadata/uns
adata.uns['dummy_category'] = pd.array(['a', 'b', None], dtype='category')

adata.uns['dummy_bool'] = [True, True, False]
adata.uns['dummy_bool2'] = pd.array([True, False, None])

adata.uns['dummy_int'] = [1,2,3]
adata.uns['dummy_int2'] = pd.array([1,2,None])

adata.write("/home/jack/src/zellkonverter/inst/extdata/krumsiek11_augmented_v0-8.h5ad")
```

lazappi commented on R/read.R:
```diff
@@ -289,16 +298,57 @@ readH5AD <- function(file, X_name = NULL, use_hdf5 = FALSE,
+
+.convert_special_types <- function(obj, recursive=FALSE) {
+    if (identical(names(obj), c('categories', 'codes'))) {
+        codes <- obj[['codes']] + 1
```
We usually use double quotes, a quick update for consistency would be great.

jackkamm replied:
Converted single to double quotes

lazappi commented on R/read.R:
```diff
@@ -289,16 +298,57 @@ readH5AD <- function(file, X_name = NULL, use_hdf5 = FALSE,
-            rhdf5::h5read(file, file.path(path, col_name))
-        )
+        out_cols[[col_name]] <- .read_convert(file, file.path(path, col_name),
+                                              recursive=FALSE)
```
Ah, I see we use `recursive=FALSE` here, can probably ignore the earlier comment then

lazappi commented on tests/testthat/test-read.R:
```diff
@@ -74,6 +75,61 @@ test_that("Reading H5AD works with native reader", {
+    expect_identical(rowData(sce_py), rowData(sce_r))
+
+    # colData colnames not in same order, but check the sets are the same
+    expect_true(setequal(colnames(colData(sce_py)), colnames(colData(sce_r))))
```
Would it be possible/easy to maintain the order? I'm pretty sure the Python reader keeps it so it should be in the fine somewhere (although not sure I have properly checked this).

ivirshup replied:
There's a `"column-order"` attribute on dataframes for this (https://anndata.readthedocs.io/en/latest/fileformat-prose.html#dataframes)

jackkamm replied:
Updated to use the `column-order` attribute

lazappi commented on tests/testthat/test-read.R:
```diff
@@ -74,6 +75,61 @@ test_that("Reading H5AD works with native reader", {
+    expect_identical(rownames(sce_py), rownames(sce_r))
+    expect_identical(colnames(sce_py), colnames(sce_r))
+
+    # rowData is just an empty DataFrame here
+    expect_identical(rowData(sce_py), rowData(sce_r))
```
Maybe it's worth adding something to the rows to check this?

jackkamm replied:
Yep I added a string column to the rowData now

lazappi commented on tests/testthat/test-read.R:
```diff
@@ -74,6 +75,61 @@ test_that("Reading H5AD works with native reader", {
+    # colData colnames not in same order, but check the sets are the same
+    expect_true(setequal(colnames(colData(sce_py)), colnames(colData(sce_r))))
+
+    # check colData columns that Python reader is able to handle
+    good_coldat_columns <- c('cell_type', 'dummy_bool', 'dummy_int',
+                             'dummy_num', 'dummy_num2')
```
Is it the `NA` thing that doesn't work with the Python reader or something else?

lazappi commented on tests/testthat/test-read.R:
```diff
@@ -74,6 +75,61 @@ test_that("Reading H5AD works with native reader", {
+
+    # Manually check colData columns that Python reader doesn't handle
+    # right (it reads them as environment objects)
+    expect_equal(colData(sce_r)$dummy_bool2,
+                 c(FALSE, NA, rep(TRUE, 638)))
```
Ah, I see. This shouldn't be happening, any thoughts on what might be going on here?

lazappi commented on tests/testthat/test-read.R:
```diff
@@ -74,6 +75,61 @@ test_that("Reading H5AD works with native reader", {
+    # a bug in the python reader?)
+    expect_equal(
+        as.vector(metadata(sce_py)[['dummy_bool']]),
+        metadata(sce_r)[['dummy_bool']]
```
Hmmmm...not sure. It may be something to do with how **{reticulate}** converts single values.

lazappi reviewed (changes requested):
Overall I'm pretty happy with this. I have made a few comments but they are mostly questions. I think this is a fairly significant contribution so I would be happy for you to add yourself as a contributor in the `DESCRIPTION` if you like.

@LTLA @ivirshup do you have any further comments?

lazappi commented:
> I've added handling for nullable ints and bools, so I believe the implementation satisfies the v0.8 spec now [1], and is ready for a review.
> 
> For the test, I found that `writeH5AD()` wasn't quite consistent with the spec [2], therefore I manually created a separate AnnData v0.8 object in Python [3], which I saved to the git repo.

I think we need to look into what is going on with the Python reader, I'm not sure about the environment business... That can probably be a separate PR though, but if you want to open an issue about it that would be helpful.

> Footnotes:
> 
> [1] The spec also mentions string handling, but I didn't test it because AnnData.write always converts strings to factors: [scverse/anndata@`8e793af`/anndata/_io/h5ad.py#L58](https://github.com/scverse/anndata/blob/8e793af01a77d0e31e91a72f3988df7d6de9cdc5/anndata/_io/h5ad.py#L58)

I don't think it always happens, but yes, most of the time it does.

> [2] `writeH5AD` seems to convert `NA_integer_` to `-2^31`, instead of using a mask as in the spec.

Another thing to check...

> [3] Here is the Python code I used to create `krumsiek11_augmented_v0-8.h5ad`: [gist.github.com/jackkamm/3b606d15d83063ed8e5f03ae1c7ab928](https://gist.github.com/jackkamm/3b606d15d83063ed8e5f03ae1c7ab928)

This should be added to `inst/scripts` as a description of the dataset. If you could add a comment with the important package versions that would be great as well.

ivirshup commented on R/read.R:
```diff
@@ -289,16 +298,57 @@ readH5AD <- function(file, X_name = NULL, use_hdf5 = FALSE,
+}
+
+.convert_special_types <- function(obj, recursive=FALSE) {
+    if (identical(names(obj), c('categories', 'codes'))) {
```
I would strongly suggest using the attributes specifying the encoding to figure out how to read types rather than inferring from the contents. But maybe this is out of scope for this PR.

https://anndata.readthedocs.io/en/latest/fileformat-prose.html#elements

jackkamm replied:
I think it's doable by passing the path here and using `rhdf5::h5_readAttributes()` -- will look into it

jackkamm replied:
Renamed to `.convert_element`, added some comments, and used `rhdf5::h5readAttributes` to read the `encoding-type` attribute (passing in the path as well).

jackkamm commented:
Thanks for the helpful reviews :) I might be a little slow getting to it due to some other deadlines right now, but hopefully should respond & have this done later next week

jackkamm commented on R/read.R:
```diff
@@ -236,6 +239,12 @@ readH5AD <- function(file, X_name = NULL, use_hdf5 = FALSE,
+    if (("X_name" %in% names(metadata(sce))) && ("X" %in% names(contents))) {
+        stopifnot(names(assays)[1] == "X") #should be true b/c X is read 1st
+        names(assays)[1] <- metadata(sce)[["X_name"]]
+        metadata(sce)[["X_name"]] <- NULL
```
Oops not sure how that slipped through. Fixed to `assays(sce)`, and also added '@importFrom SummarizedExperiment assays assays<-`

jackkamm commented:
I've revised this PR now based on the feedback. I also squashed all the commits and force pushed.

Biggest changes are:
- Use the `encoding-type` attribute during conversion. Also pass in the path to the recursive conversion function in order to do this.
- Pass `version` along to internal functions as @LTLA recommended earlier. If version < 0.8, then behavior of native reader is unchanged from before.

Also in regard to this:

> I don't think it [string to factor conversion] always happens, but yes, most of the time it does.

Seems like the conversion doesn't happen when the string values are all unique. I added such a column of unique strings to the test data. rhdf5 seems able to read it without any issue.

jackkamm commented:
Another minor comment I forgot to add:

>  [2] writeH5AD seems to convert NA_integer_ to -2^31, instead of using a mask as in the spec.
> Another thing to check...

This might be a relatively minor issue. When I tested it before, rhdf5 and h5py both converted the -2^31 to NA/nan when reading it into R/python.

lazappi commented on R/read.R:
```diff
@@ -99,22 +99,24 @@ readH5AD <- function(file, X_name = NULL, use_hdf5 = FALSE,
     .ui_info("Using the {.field R} reader")
     .ui_step("Reading {.file {file}}", spinner = TRUE)
 
+    version <- match.arg(version, .AnnDataVersions)
+
```
I'm not sure if this is the right thing to do. The current `.AnnDataVersions` records the versions there are **{basilisk}** environments for but that might not match up with all the possible Python versions. It probably depends a bit what this is used for.

jackkamm replied:
Agreed. I changed it now to:
```
if (is.null(version)) {
    version <- .AnnDataVersions[1]
}
```
so when it's null it'll match the default python version, but otherwise isn't constrained.

lazappi commented on R/read.R:
```diff
@@ -289,23 +309,89 @@ readH5AD <- function(file, X_name = NULL, use_hdf5 = FALSE,
 }
 
+.convert_element <- function(obj, path, file, recursive=FALSE) {
+    # Should we wrap this in suppressWarnings? rhdf5 will warn that it
+    # can't yet read enum (factor/boolean) in attributes
+    element_attrs <- rhdf5::h5readAttributes(file, path)
```
Hmmm...maybe. If it's something we handle so the warning isn't meaningful to the user then I guess that would make sense. There's just the risk that we suppress a future warning that it would be useful to know about.

lazappi commented on R/read.R:
```diff
@@ -289,23 +309,89 @@ readH5AD <- function(file, X_name = NULL, use_hdf5 = FALSE,
+        codes[codes == 0] <- NA
+        levels <- obj[["categories"]]
+
+        # Can't determine orderedness due to rhdf5 not yet supporting
+        # enums in attributes
+        #ord <- as.logical(element_attrs[["ordered"]])
```
Ah, I guess this is where the warning comes in. If we aren't really addressing it maybe we should leave it (and try to fix this upstream).

jackkamm replied:
Agree it would be good to fix this upstream...

Also, I consolidated the 2 comments into a single comment, to try and make this issue easier to keep track of.

jackkamm commented on R/read.R:
```diff
@@ -289,23 +311,87 @@ readH5AD <- function(file, X_name = NULL, use_hdf5 = FALSE,
+            )
+            out_cols[[cat_name]] <- factor(out_cols[[cat_name]])
+            levels(out_cols[[cat_name]]) <- levels
+        }
```
This is the problem I mentioned in my comment just now. AnnData 0.7 categories are only converted if `version` explicitly specified.

lazappi commented:
So, in the Python reader the version is there to control which environment is used (and therefore which AnnData file version is read/written). In the R reader we aren't messing around with environments so it maybe isn't needed. It just depends whether it's easier to detect what file version has been used or ask the user to specify it.

If we were looking at a writer it would be a bit different because in that case we would want to give the user control over which file version is written.

jackkamm commented:
I made 1 more commit, so that the version isn't passed into native R reader anymore. Instead, native R reader just tries to figure out the right thing to do based on the keywords/attributes it sees, like in the original version of this PR.

I think it's probably better this way, so the user doesn't need to explicitly specify the version.

But I can see the argument the other way also. So I leave this last change as a separate unsquashed commit -- feel free to revert it if you prefer the previous approach that explicitly passes in the version.

lazappi commented:
Thanks for all your work on this! I suspect it might need some tweaks in the future but I have merged what we have so far.
